types/wlr_pointer_gestures_v1.c

```diff
@@ -0,0 +1,274 @@
+		uint32_t fingers) {
+	struct wl_resource *pinch;
+	wl_resource_for_each(pinch, &gestures->pinches) {
+		struct wl_client *pinch_client = wl_resource_get_client(pinch);
+		if (pinch_client != wl_resource_get_client(focus->resource)) {
+			continue;
+		}
+		uint32_t serial = wl_display_next_serial(
+				wl_client_get_display(pinch_client));
+		zwp_pointer_gesture_pinch_v1_send_begin(pinch, serial,
+				time_msec, focus->resource, fingers);
+	}
+}
+
+void wlr_pointer_gesture_pinch_notify_update(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_surface *focus,
+		uint32_t time_msec,
+		double dx,
+		double dy,
+		double scale,
+		double rotation) {
+	struct wl_resource *pinch;
+	wl_resource_for_each(pinch, &gestures->pinches) {
+		struct wl_client *pinch_client = wl_resource_get_client(pinch);
+		if (pinch_client != wl_resource_get_client(focus->resource)) {
+			continue;
+		}
+		zwp_pointer_gesture_pinch_v1_send_update(pinch, time_msec,
+				wl_fixed_from_double(dx), wl_fixed_from_double(dy),
+				wl_fixed_from_double(scale),
+				wl_fixed_from_double(rotation));
+	}
+}
+
+void wlr_pointer_gesture_pinch_notify_end(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_surface *focus,
+		uint32_t time_msec,
+		bool cancelled) {
+	struct wl_resource *pinch;
+	wl_resource_for_each(pinch, &gestures->pinches) {
+		struct wl_client *pinch_client = wl_resource_get_client(pinch);
+		if (pinch_client != wl_resource_get_client(focus->resource)) {
+			continue;
+		}
+		uint32_t serial = wl_display_next_serial(
+				wl_client_get_display(pinch_client));
+		zwp_pointer_gesture_pinch_v1_send_end(pinch, serial,
+				time_msec, cancelled);
+	}
+}
+
+static const struct zwp_pointer_gesture_pinch_v1_interface pinch_impl = {
+	.destroy = resource_handle_destroy,
+};
+
+static void get_pinch_gesture(struct wl_client *client,
+		struct wl_resource *gestures_resource,
+		uint32_t id,
+		struct wl_resource *pointer_resource) {
+	struct wlr_pointer_gestures_v1 *gestures =
+		pointer_gestures_from_resource(gestures_resource);
+
+	struct wl_resource *pinch = wl_resource_create(client,
+		&zwp_pointer_gesture_pinch_v1_interface,
+		wl_resource_get_version(gestures_resource),
+		id);
+	if (pinch == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(pinch, &pinch_impl, NULL, resource_remove_from_list);
+	wl_list_insert(&gestures->pinches, wl_resource_get_link(pinch));
+}
+
+static const struct zwp_pointer_gestures_v1_interface gestures_impl = {
+	.get_swipe_gesture = get_swipe_gesture,
+	.get_pinch_gesture = get_pinch_gesture,
+};
+
+static void client_handle_destroy(struct wl_resource *resource) {
+	struct wl_client *client = wl_resource_get_client(resource);
+	struct wlr_pointer_gestures_v1 *gestures =
+		pointer_gestures_from_resource(resource);
+	struct wl_resource *gesture, *tmp = NULL;
+	wl_resource_for_each_safe(gesture, tmp, &gestures->swipes) {
+		if (wl_resource_get_client(gesture) == client) {
```

Comment:
This doesn't work if the client binds twice to the gestures global. Moreover, the protocol doesn't say swipe and pinch objects become invalid when the gestures object is destroyed. I think we shouldn't destroy these here.

Reply:
All of the client's resources are destroyed (with destructors run) when it disconnects, right?

Reply:
Yes. The order of destruction isn't guaranteed. Also, there is one other situation when the resource destructor is called: when the client sends the `destroy` request on an object. So the client can e.g. create bind twice to the same global, then destroy one.